File: lambdas/utils/errorHandler.js
/**
 * Error Handler Utility
 * Centralized error handling for Lambda functions
 */

const response = require('./response');

/**
 * Handle errors and return appropriate response
 */
const handleError = (error) => {
  console.error('Error:', error);

  // Database errors
  if (error.code === 'ER_DUP_ENTRY') {
    return response.error('Duplicate entry. Record already exists.', 409);
  }

  if (error.code === 'ER_NO_REFERENCED_ROW_2') {
    return response.error('Referenced record does not exist.', 400);
  }

  // JWT/Auth errors
  if (error.message && error.message.includes('token')) {
    return response.unauthorized(error.message);
  }

  // Validation errors
  if (error.name === 'ValidationError') {
    return response.validationError(error.details);
  }

  // Default error
  return response.error(
    error.message || 'An unexpected error occurred',
    error.statusCode || 500
  );
};

/**
 * Wrapper function to handle async errors
 */
const asyncHandler = (fn) => {
  return async (event, context) => {
    try {
      return await fn(event, context);
    } catch (error) {
      return handleError(error);
    }
  };
};

module.exports = {
  handleError,
  asyncHandler
};

File: lambdas/utils/error_handler.py
"""
Error Handler Utility
Centralized error handling for Lambda functions
"""

from . import response
import traceback


def handle_error(error):
    """Handle errors and return appropriate response"""
    print(f'Error: {error}')
    traceback.print_exc()
    
    error_str = str(error)
    
    # Database errors
    if 'Duplicate entry' in error_str:
        return response.error('Duplicate entry. Record already exists.', 409)
    
    if 'foreign key constraint' in error_str.lower():
        return response.error('Referenced record does not exist.', 400)
    
    # JWT/Auth errors
    if 'token' in error_str.lower() or 'authorization' in error_str.lower():
        return response.unauthorized(error_str)
    
    # Validation errors
    if hasattr(error, 'name') and error.name == 'ValidationError':
        return response.validation_error(getattr(error, 'details', None))
    
    # Default error
    status_code = getattr(error, 'status_code', 500)
    return response.error(error_str or 'An unexpected error occurred', status_code)


def async_handler(func):
    """Decorator to handle async errors"""
    def wrapper(event, context):
        try:
            return func(event, context)
        except Exception as error:
            return handle_error(error)
    return wrapper
